modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -0,0 +1,312 @@
+
+(defn- ^BigQuery database->client
+  [database]
+  (let [creds   (database->service-account-credential database)
+        bq-bldr (doto (BigQueryOptions/newBuilder)
+                  (.setCredentials (.createScoped creds (Collections/singletonList bigquery-scope))))]
+    (.. bq-bldr build getService)))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                      Sync                                                      |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defn- ^Iterable list-tables
+  "Fetch all tables (new pages are loaded automatically by the API)."
+  ([{{:keys [project-id dataset-id]} :details, :as database}]
+   (let [bq (database->client database)]
+     (list-tables bq project-id dataset-id)))
+
+  #_([{{:keys [project-id dataset-id]} :details, :as database}, ^String page-token-or-nil]
+     ;; RAR 2020-May-11 - this gets messy, because this function is used to determine if we
+     ;; can connect to the database. If a service account JSON is being used, the project ID
+     ;; field isn't required in the admin page because it's embedded in that JSON. During the
+     ;; initial save, the database entry doesn't exist yet, so we haven't pulled the project ID
+     ;; from that JSON. In this case, we need to go look it up in the credentials.
+     ;; TODO: maybe handle this better in `driver/can-connect? :bigquery-cloud-sdk`
+     (let [bq (database->client database)]
+      (list-tables db-client project-id dataset-id page-token-or-nil)))
+
+  ([^BigQuery client, ^String project-id ^String dataset-id]
+   {:pre [client (not (str/blank? dataset-id))]}
+   (.iterateAll (.listTables client (DatasetId/of project-id dataset-id) (u/varargs BigQuery$TableListOption [])))))
+
+(defmethod driver/describe-database :bigquery-cloud-sdk
+  [_ database]
+  ;; first page through all the 50-table pages until we stop getting "next page tokens"
+  (let [tables (list-tables database) #_(loop [tables [], ^TableList table-list (list-tables database)]
+                                          (let [tables (concat tables (.getTables table-list))]
+                                            (if-let [next-page-token (.getNextPageToken table-list)]
+                                              (recur tables (list-tables database next-page-token))
+                                              tables)))]
+    ;; after that convert the results to MB format
+    {:tables (set (for [^Table table tables]
+                    {:schema nil, :name (.. table getTableId getTable)}))}))
+
+(defmethod driver/can-connect? :bigquery-cloud-sdk
+  [_ details-map]
+  ;; check whether we can connect by seeing whether we have at least one Table in the iterator
+  (let [^Iterable table-iter (list-tables {:details details-map})]
+    (.. table-iter iterator hasNext)))
```

Comment:
can this distinguish between connecting to a dataset that has no tables in it versus unable to connect to a dataset?

Reply:
Yes; the latter case will throw an exception (earlier on, actually, in the `database->client` call).

Reply:
but won't `hasNext` return false on a dataset with no tables, causing `driver/can-connect?` to return false to a dataset it can connect to which has no tables?

Reply:
```clojure
(.. (list-tables {:details {:project-id "metabase-testing-311615"
                            :service-account-json "redacted"
                            :dataset-id "empty"
                            :engine :bigquery-cloud-sdk}})
    iterator hasNext)
```

this returns false, even though it connects just fine and does not throw an error. It was a new dataset i created `bq mk empty` that has no tables in it. It seems like we would more likely want something like

`(try (list-tables db) true (catch Exception e false))` if list-tables will throw if it cannot connect.

Reply:
Ah I see your point.  Yes, I will fix that.

Reply:
OK, this has been fixed now.  `can-connect?` will instead simply check for the presence of the given `dataset-id` (not even looking at tables).  Added a test while I'm at it.